examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3181 @@
+//! the already deployed contract addresses into the proxy `init` function.
+//! Then, call the `initialize` functions on the `state` as well as on the
+//! `implementation` contracts to set the remaining addresses.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use concordium_cis2::*;
+use concordium_std::{fmt::Debug, *};
+
+/// The id of the wCCD token in this contract.
+const TOKEN_ID_WCCD: ContractTokenId = TokenIdUnit();
+
+/// The metadata url for the wCCD token.
+const TOKEN_METADATA_URL: &str = "https://some.example/token/wccd";
+
+// Types
+
+/// Contract token ID type.
+/// Since this contract will only ever contain this one token type, we use the
+/// smallest possible token ID.
+type ContractTokenId = TokenIdUnit;
+
+/// Contract token amount type.
+/// Since this contract is wrapping the CCD and the CCD can be represented as a
+/// u64, we can specialize the token amount to u64 reducing module size and cost
+/// of arithmetics.
+type ContractTokenAmount = TokenAmountU64;
+
+/// Needed for the custom serial instance, which doesn't include the `Option`
+/// tag and the length of the vector.
+#[derive(PartialEq, Eq, Debug)]
+struct RawReturnValue(Option<Vec<u8>>);
+
+impl Serial for RawReturnValue {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        match &self.0 {
+            Some(rv) => out.write_all(rv),
+            None => Ok(()),
+        }
+    }
+}
+
+/// The state tracked for each address.
+#[derive(Serial, DeserialWithState, Deletable)]
+#[concordium(state_parameter = "S")]
+struct AddressState<S> {
+    /// The number of tokens owned by this address.
+    balance:   ContractTokenAmount,
+    /// The address which are currently enabled as operators for this token and
+    /// this address.
+    operators: StateSet<Address, S>,
+}
+
+/// The `state` contract state.
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// Address of the w_ccd proxy contract.
+    proxy_address:          Option<ContractAddress>,
+    /// Address of the w_ccd implementation contract.
+    implementation_address: Option<ContractAddress>,
+    /// The state of the one token.
+    token:                  StateMap<Address, AddressState<S>, S>,
+    /// Timestamp when contract is unpaused.
+    unpause_time:           Timestamp,
+}
+
+/// The `implementation` contract state.
+#[derive(Serial, Deserial, SchemaType)]
+struct StateImplementation {
+    /// The admin address can pause/unpause the contract.
+    admin:          Address,
+    /// Address of the w_ccd proxy contract.
+    proxy_address:  Option<ContractAddress>,
+    /// Address of the w_ccd state contract.
+    state_address:  Option<ContractAddress>,
+    /// Fallback function on the proxy will set this value
+    /// before invoking a function on the implementation.
+    current_sender: Address,
+}
+
+/// The `proxy` contract state.
+#[derive(Serial, Deserial, SchemaType)]
+struct StateProxy {
```

Comment:
I tried it out but removing it will cause an error during the schema generation. `StateImplementation` and `StateProxy` are used as return values for the `view` function.

Reply:
Ok. In the ideal world that would be a separate type, but in this case that does not seem the most sensible option. So we keep it :)